x/gamm/keeper/swap.go

```diff
@@ -45,35 +45,59 @@ func (k Keeper) swapExactAmountIn(
 	tokenOutMinAmount sdk.Int,
 	swapFee sdk.Dec,
 ) (tokenOutAmount sdk.Int, err error) {
+	pool, tokenOut, err := k.swapExactAmountInNoTokenSend(ctx, pool, tokenIn, tokenOutDenom, tokenOutMinAmount, swapFee)
+	if err != nil {
+		return sdk.Int{}, err
+	}
+
+	// Settles balances between the tx sender and the pool to match the swap that was executed earlier.
+	// Also emits swap event and updates related liquidity metrics
+	if err := k.updatePoolForSwap(ctx, pool, tokenIn, tokenOut); err != nil {
+		return sdk.Int{}, err
+	}
+
+	err = k.sendCoinsAfterSwap(ctx, pool, tokenIn, tokenOut, sender)
+	if err != nil {
+		return sdk.Int{}, err
+	}
+
+	return tokenOut.Amount, nil
+}
+
+// swapExactAmountInNoTokenSend performs `SwapOutAmtGivenIn` for the given inputs,
+// and potentially mutates state. It does not save the new pool struct, or do token transfers.
+// Returns the altered pool after swap, and the token out when swapped.
+func (k Keeper) swapExactAmountInNoTokenSend(
+	ctx sdk.Context,
+	pool types.PoolI,
+	tokenIn sdk.Coin,
+	tokenOutDenom string,
+	tokenOutMinAmount sdk.Int,
+	swapFee sdk.Dec,
+) (updatedPool types.PoolI, tokenOut sdk.Coin, err error) {
```

Comment:
Do we even need to return `updatedPool` here? `pool` is a pointer so it should be mutated if that's desired

Reply:
ur right, they do get mutated if desired, removing returning the updatedPool